lib/haskell/anyall/src/AnyAll/BoolStructTree.hs

```diff
@@ -49,4 +49,30 @@ instance Monoid lbl => Semigroup (BoolStructDT lbl a) where
   (Node (FAll x) xs)  <> (Node (FAll y) ys) = Node (FAll (x<>y)) (xs <> ys)
   l                   <> (Node (FAll y) ys) = Node (FAll y) (l:ys)
   l@(Node (FAll _) _) <> r          = r <> l
-  l                   <> r          = Node (FAll mempty) [l, r]
\ No newline at end of file
+  l                   <> r          = Node (FAll mempty) [l, r]
+
+simplifyItemDT :: (Eq lbl, Monoid lbl) => BoolStructDT lbl a -> BoolStructDT lbl a
+simplifyItemDT orig = orig
+
+data MergeResult a = Merged a | Unmerged a a
+
+attemptMergeHeads :: Eq lbl => BoolStructDT lbl a -> BoolStructDT lbl a -> MergeResult (BoolStructDT lbl a)
+attemptMergeHeads  x@(Node (FAll xl) xs)  y@(Node (FAll yl) ys)
+  | xl == yl = Merged $ Node (FAll xl) (xs++ys)
+  | otherwise = Unmerged x y
+attemptMergeHeads  x@(Node (FAny xl) xs)  y@(Node (FAny yl) ys)
+  | xl == yl = Merged $ Node (FAny xl) (xs++ys)
+  | otherwise = Unmerged x y
+attemptMergeHeads  x  y = Unmerged x y
+
+mergeMatch :: (Eq lbl, Monoid lbl) => [BoolStructDT lbl a] -> [BoolStructDT lbl a]
+mergeMatch []  = []
+mergeMatch [k] = [k]
+mergeMatch (bs1 : bs2 : zs) = case x of
+  (Merged m) -> mergeMatch (m:zs)
+  (Unmerged x y) -> x : mergeMatch (y:zs)
```

Comment:
does the linter give a "shadows existing binding" error for this reuse of `x`?

Reply:
No, linter from VS Code doesn't.
Which linter are you using?